openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -1,144 +1,203 @@
+            args - The list of positional arguments supplied when the decorated function was called.
+            kwargs - The dictionary of keyword arguments supplied when the decorated function was called.
 
-        Args:
-            function - wrapped function i.e. real_donottrack
         """
-        def wrapper(*args, **kwargs):
-            """ wrapper function for decorated function
-            Returns:
-                wrapper function i.e. wrapper
-            """
-            if len(classes_not_to_be_tracked) == 0:
-                global TRACK_FLAG  # pylint: disable=W0603
-                current_flag = TRACK_FLAG
-                TRACK_FLAG = False
-                function(*args, **kwargs)
-                TRACK_FLAG = current_flag
+        if entities_not_to_be_tracked:
+            global HALT_TRACKING  # pylint: disable=W0603
+            HALT_TRACKING.append(entities_not_to_be_tracked)
+            HALT_TRACKING[-1] = list(set([x for sublist in HALT_TRACKING for x in sublist]))
+            return_value = wrapped(*args, **kwargs)
+
+            # check if the returning class is a generator
+            if isinstance(return_value, types.GeneratorType):
+                def generator_wrapper(wrapped_generator):
+                    try:
+                        while True:
+                            return_value = next(wrapped_generator)
+                            yield return_value
+                    finally:
+                        global HALT_TRACKING
+                        HALT_TRACKING.pop()
+                return generator_wrapper(return_value)
+            else:
+                HALT_TRACKING.pop()
+                return return_value
+
+        else:  # if donottrack is not parameterized
+            global TRACK_FLAG  # pylint: disable=W0603
+            TRACK_FLAG = False
+            return_value = wrapped(*args, **kwargs)
+
+            # check if the returning class is a generator
+            if isinstance(return_value, types.GeneratorType):
+                def generator_wrapper(wrapped_generator):
+                    try:
+                        while True:
+                            return_value = next(wrapped_generator)
+                            yield return_value
+                    finally:
+                        global TRACK_FLAG
+                        TRACK_FLAG = True
+                return generator_wrapper(return_value)
             else:
-                global HALT_TRACKING  # pylint: disable=W0603
-                current_halt_track = HALT_TRACKING
-                HALT_TRACKING = classes_not_to_be_tracked
-                function(*args, **kwargs)
-                HALT_TRACKING = current_halt_track
-        return wrapper
+                TRACK_FLAG = False
+                return return_value
     return real_donottrack
+
+
+@wrapt.decorator()
+def trackit(wrapped, instance, args, kwargs): # pylint: disable=W0613
+    """ Decorator 
```

Comment:
Please add a more thorough docstring, so that users and developers will actually understand what this decorator does.